i18n - legacy message id changes

petebacondarwin opened the pull request:
This PR implements [Common digest function](https://hackmd.io/EQF4_-atSXK4XWg8eAha2g?both#Common-digest-function) functionality of the legacy message-id support.

AndrewKushnir reviewed (approved):
LGTM 👍 Thanks Pete!

AndrewKushnir commented:
Started presubmits to get early results:
- [VE Presubmit](https://test.corp.google.com/ui#id=OCL:271472721:BASE:271472739:1569545852720:1b35cee8)
- [Ivy Presubmit](http://test/OCL:271472721:BASE:271472941:1569545926322:e2f3fb5c)

ocombe reviewed (approved):
I don't think that you should put a breaking change message in `fix(ivy): i18n - do not render message ids unnecessarily`, as it might appear in the v9 release changelog, even if you're not breaking anything that existed previously

AndrewKushnir commented:
FYI, VE and Ivy presubmits are successful. Thank you.

IgorMinar commented on packages/localize/src/utils/messages.ts:
```diff
@@ -158,35 +159,61 @@ export function parseMetadata(cooked: string, raw: string): MessageMetadata {
+ * Find the end of a "marked block" indicated by the first non-escaped colon.
+ *
+ * This function is repeated in `src/localize/src/localize.ts` and the two should be kept in sync.
+ * See that file for more explanation of why.
```
can you make this note much more prominent and move it closer to the function declaration (end of the comment block), so that it's easier to see it if the code review diff only shows a few lines of the comment block. thank you

petebacondarwin replied:
Done

IgorMinar commented on integration/cli-hello-world-ivy-i18n/src/app/app.module.ts:
```diff
@@ -1,21 +1,20 @@
-import { AppComponent } from './app.component';
+// We must load translations before `AppComponent` is imported
+// Otherwise the `$localize` calls inside its template will not be translated
+import './translations';
```
we should not be relying on the fact that application code can have side-effecty module imports because build optimizations will likely break such code. (I know that this is just a test app, but some people will copy it). Please rewrite this into code that doesn't have side-effecty imports.

petebacondarwin replied:
I'm not quite sure how I can do that without adding this to the `polyfills.ts` file.

What would be cool is if we could tell Angular CLI that there is a file that must be executed before the main bundle...

IgorMinar replied:
the main.ts file is one that must have side effects (the bootstrap call), so what I suggest is that you export `configTranslations` or `loadTranslations` fn from `translations.ts`, and then import and call it from `main.ts`.

petebacondarwin replied:
Sadly that doesn't work because the call to `loadTranslations()` (the one from `@angular/localize`) must execute before the module loader even evaluates the `@Component()` decoration of `AppComponent`.

Any code that is inside the body of `main.ts` will be evaluated after all the imports have been evaluated - whatever order they actually occur in the file - and so putting any kind of code execution (e.g. calling `configureTranslations()` etc) in main.ts will fail to run early enough.

The only way for it to work is either to have a side-effect-y import like this or to put it into a bundle that will be executed before the main bundle (i.e. polyfills).

I am going to go with the latter since that is where we already put the import of `@angular/localize/init` anyway.

IgorMinar commented on integration/cli-hello-world-ivy-i18n/src/app/translations.ts:
```diff
@@ -0,0 +1,8 @@
+import {computeMsgId} from '@angular/compiler';
```
we should add a note here that this is not a public api - nothing in the @angular/compiler entry point is a public api.

petebacondarwin replied:
Done

IgorMinar reviewed (approved):
a few small changes requested, but otherwise looks good.

petebacondarwin committed:
fix(ivy): i18n - do not render message ids unnecessarily

In an attempt to be compatible with previous translation files
the Angular compiler was generating instructions that always
included the message id. This was because it was not possible
to accurately re-generate the id from the calls to `$localize()` alone.

In line with https://hackmd.io/EQF4_-atSXK4XWg8eAha2g this
commit changes the compiler so that it only renders ids if they are
"custom" ones provided by the template author.

NOTE:

When translating messages generated by the Angular compiler
from i18n tags in templates, the `$localize.translate()` function
will compute message ids, if no custom id is provided, using a
common digest function that only relies upon the information
available in the `$localize()` calls.

This computed message id will not be the same as the message
ids stored in legacy translation files. Such files will need to be
migrated to use the new common digest function.

This only affects developers who have been trialling `$localize`, have
been calling `loadTranslations()`, and are not exclusively using custom
ids in their templates.

petebacondarwin committed:
fix(ivy): i18n - support colons in $localize metadata

Metadata blocks are delimited by colons. Previously the code naively just
looked for the next colon in the string as the end marker.

This commit supports escaping colons within the metadata content.
The Angular compiler has been updated to add escaping as required.

petebacondarwin committed:
test(ivy): add runtime translations to integration test app

petebacondarwin committed:
fix(ivy): i18n - throw an error if a translation contains an invalid placeholder

Previously if a translation contains a placeholder that
does not exist in the message being translated, that
placeholder is evaluated as `undefined`.

Translations should never contain such placeholder names
so now `translate` will throw a helpful error in instead.

petebacondarwin committed:
refactor(ivy): i18n - move marker block serialization to helpers

Previously the metadata and placeholder blocks were serialized in
a variety of places. Moreover the code for creating the `LocalizedString`
AST node was doing serialization, which break the separation of concerns.

Now this is all done by the code that renders the AST and is refactored into
helper functions to avoid repeating the behaviour.

petebacondarwin committed:
fix(ivy): i18n - better translation warnings

The missing translation and invalid placeholder warnings now contain the
"meaning" if one was provided.
